project3/coordinator/message_log.h

```diff
@@ -0,0 +1,70 @@
+/**
+ * @file The coordinator's message log.
+ */
+#ifndef PROJECT3_MESSAGE_LOG_H
+#define PROJECT3_MESSAGE_LOG_H
+#include <algorithm>
+#include <mutex>
+#include <unordered_set>
+#include <memory>
+namespace coordinator::message_log {
+
+/**
+ * @class A log of all messages broadcast by the coordinator.
+ */
+class MessageLog {
+ public:
+  /**
+   * Structure of messages stored by the coordinator.
+   */
+  struct Message {
+    bool operator==(const Message&) const;
+    /// The actual message.
+    std::string msg;
+
+    /// The time the message was sent.
+    double timestamp;
```

Comment:
I think storing the timestamp as a `double` here is going to be sub-optimal. For one, you're relying on this for your equality checks and floating-point values don't always compare nicely to each-other. Furthermore, C++ has the [`std::chrono::timepoint`](https://en.cppreference.com/w/cpp/chrono/time_point) class which I think suits your needs better.